src/EditorFeatures/Core.Wpf/Suggestions/AsyncSuggestedActionsSource.cs

```diff
@@ -111,41 +111,123 @@ private partial class AsyncSuggestedActionsSource : SuggestedActionsSource, IAsy
-
                         if (priority != null)
                         {
+                            priorityProvider = priorityProvider.With(priority.Value);
+
+                            // Compute the actions sets for the current request priority.
                             var allSets = GetCodeFixesAndRefactoringsAsync(
                                 state, requestedActionCategories, document,
                                 range, selection,
                                 addOperationScope: _ => null,
-                                priority.Value,
+                                priorityProvider,
                                 currentActionCount, cancellationToken).WithCancellation(cancellationToken).ConfigureAwait(false);
 
+                            // Add those computed actions sets to the current collector set which they have
+                            // matching or higher priority compared to the request priority.
+                            // For computed action sets with a lower priority compared to the request priority,
+                            // add them to the appropriate pending set so it gets added to the collector set in later iteration.
                             await foreach (var set in allSets)
                             {
-                                if (priority == CodeActionRequestPriority.High && set.Priority == SuggestedActionSetPriority.Low)
+                                switch (priority)
                                 {
-                                    // if we're processing the high pri bucket, but we get action sets for lower pri
-                                    // groups, then keep track of them and add them in later when we get to that group.
-                                    lowPrioritySets.Add(set);
-                                }
-                                else
-                                {
-                                    currentActionCount += set.Actions.Count();
-                                    collector.Add(set);
+                                    case CodeActionRequestPriority.High:
+                                        switch (set.Priority)
+                                        {
+                                            case SuggestedActionSetPriority.High:
+                                                break;
+
+                                            case SuggestedActionSetPriority.Medium:
+                                                pendingMediumPrioritySets.Add(set);
+                                                continue;
+
+                                            case SuggestedActionSetPriority.Low:
+                                                pendingLowPrioritySets.Add(set);
+                                                continue;
+
+                                            default:
+                                                throw ExceptionUtilities.UnexpectedValue(set.Priority);
```

Comment:
nit: consider pulling this check up higher.  all your branches need it.  so it would clean things up.

Reply:
Going to revert